src/components/AnnotationLayer.tsx

```diff
@@ -510,98 +540,124 @@ const createAnnotations = (
-    }
+export default (props: AnnotationLayerProps) => {
+  const { legendSettings, margin, size, annotations, annotationHandling } =
+    props
 
-    this.state = {
-      ...baseState,
-      ...createAnnotations(props, baseState)
-    }
-  }
+  const [SpanOrDiv] = useState(() => HOCSpanOrDiv(props.useSpans))
 
-  static getDerivedStateFromProps(
-    nextProps: AnnotationLayerProps,
-    prevState: AnnotationLayerState
-  ) {
-    return createAnnotations(nextProps, prevState)
-  }
+  const annotationProcessor: AnnotationHandling =
+    typeof annotationHandling === "object"
+      ? annotationHandling
+      : { layout: { type: annotationHandling }, dataVersion: "" }
+
+  const { dataVersion = "" } = annotationProcessor
 
-  render() {
-    const { svgAnnotations, htmlAnnotations, SpanOrDiv } = this.state
-    const { legendSettings, margin, size } = this.props
+  const [adjustedAnnotations, changeAdjustedAnnotations] = useState([])
+  const [svgAnnotations, changeSVGAnnotations] = useState([])
+  const [htmlAnnotations, changeHTMLAnnotations] = useState([])
+  const [adjustedAnnotationsKey, changeAdjustedAnnotationsKey] = useState("")
+  const [adjustedAnnotationsDataVersion, changeAdjustedAnnotationsDataVersion] =
+    useState(dataVersion)
+
+  const initialSVGAnnotations: NoteType[] = generateSVGAnnotations(
+    props,
+    annotations
+  )
+
+  const adjustableAnnotations = initialSVGAnnotations.filter(
+    (d) => d.props && d.props.noteData && !d.props.noteData.fixedPosition
+  )
+
+  const updatedAnnotationsKey = keyFromSVGAnnotations(
+    adjustableAnnotations,
+    annotationProcessor,
+    size
+  )
+
+  useEffect(() => {
+    const fixedAnnotations = initialSVGAnnotations.filter(
+      (d) => !d.props || !d.props.noteData || d.props.noteData.fixedPosition
+    )
 
-    let renderedLegend
-    if (legendSettings) {
-      const positionHash = {
-        left: [15, 15],
-        right: [size[0] + 15, 15]
-      }
-      const { position = "right", title = "Legend" } = legendSettings
-      const legendPosition = positionHash[position]
-      renderedLegend = (
-        <g transform={`translate(${legendPosition.join(",")})`}>
-          <Legend {...legendSettings} title={title} position={position} />
-        </g>
-      )
+    const updatedState = createAnnotations(props, {
+      adjustedAnnotations,
+      adjustedAnnotationsKey,
+      adjustedAnnotationsDataVersion,
+      adjustableAnnotations,
+      fixedAnnotations
+    })
+
+    changeAdjustedAnnotations(updatedState.adjustedAnnotations)
+    changeAdjustedAnnotationsKey(updatedState.adjustedAnnotationsKey)
+    changeAdjustedAnnotationsDataVersion(
+      updatedState.adjustedAnnotationsDataVersion
+    )
+    changeSVGAnnotations(updatedState.svgAnnotations)
+    changeHTMLAnnotations(updatedState.htmlAnnotations)
+  }, [
+    updatedAnnotationsKey,
+    dataVersion,
+    annotations.length,
+    annotations.map((a) => safeStringify(a)).join("-")
+  ])
```

Comment:
Given this effect only does computations, and the fact that part of those computations happening before effect, it might be easier to just do a single big `useMemo()` to compute all the stuff related to `annotations`

Reply:
It doesn't just do computations, it also stores the prior adjusted annotations to ensure that if the existing annotations have been adjusted it shouldn't re-adjust them. I'm sure it can be optimized later but for now I'll leave it as a hook version of the logic previously in componentDidUpdate